pkg/receive/handler.go

```diff
@@ -629,6 +599,77 @@ func isNotReady(err error) bool {
 		status.Code(err) == codes.Unavailable
 }
 
+// countCause counts the number of errors within the given error slice
+// whose causes satisfy the given function.
+// cause will inspect the error's cause or, if the error is a MultiError,
+// the cause of each contained error but will not traverse any deeper.
+func countCause(errs []error, f func(error) bool) int {
+	var n int
+	for _, err := range errs {
+		if f(errors.Cause(err)) {
+			n++
+		}
+	}
+	return n
+}
+
+type expectedErrors []*struct {
+	err   error
+	cause func(error) bool
+	count int
+}
+
+func (a expectedErrors) Len() int           { return len(a) }
+func (a expectedErrors) Swap(i, j int)      { a[i], a[j] = a[j], a[i] }
+func (a expectedErrors) Less(i, j int) bool { return a[i].count < a[j].count }
+
+// cause returns a known error which has occurred more than the given threshold.
+// cause will inspect the error's cause or, if the error is a MultiError,
+// the cause of each contained error but will not traverse any deeper.
+func cause(err error, threshold int) error {
```

Comment:
I think we can work on this name ... I think it could be more verbose. It means it can masks error, right? and you use it in wrap which can be confusing a lot...